### PR TITLE
Bug in loc did not raise KeyError when missing combination with slice(None) was given

### LINKED ISSUE #19556: Inconsistent MultiIndex behaviour when indexing selects no rows
```python
mi = pd.MultiIndex.from_arrays([
        np.array(['a', 'a', 'b', 'b']), 
        np.array(['1', '2', '2', '3']), 
        np.array(['alpha', 'beta', 'alpha', 'beta']),
    ], names=['one', 'two', 'three'])
df = pd.DataFrame(np.random.rand(4, 3), index=mi)
df2 = df.loc[('b', '1', slice(None)), :]
print(df2)
print(df2.index)
```
Output
```
Empty DataFrame
Columns: [0, 1, 2]
Index: []
MultiIndex(levels=[['a', 'b'], ['1', '2', '3'], ['alpha', 'beta']],
           labels=[[], [], []],
           names=['one', 'two', 'three'])
```
Indexing `df` results in an empty dataframe in this case. However, if the third level of the MultiIndex is left out (as below), it results in a `KeyError`.
```python
df2 = df.loc[('b', '1'), :]
```

In my opinion, this behaviour is confusing and inconsistent. Preferably, both cases would throw `KeyError`.

This particular behaviour was previously discussed in issue #19543.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-43-Microsoft
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: None
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #19556
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Additional question: When an non existent key is given for ``get_locs`` it raises a ``KeyError``. When just a missing combination of keys is given (``(b,1)`` in the new test), while the individual keys exist, an empty array is returned. Is this the expected behavior?

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1107,6 +1107,8 @@ def _getitem_axis(self, key, axis: int):
             # nested tuple slicing
             if is_nested_tuple(key, labels):
                 locs = labels.get_locs(key)
+                if not len(locs):
+                    raise KeyError(key)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:1111]: if this is going to raise (not 100% sure, but leaning towards yes), i think it should happen in MultiIndex.get_locs

Comment (phofl) [reply pandas/core/indexing.py:1111]: Could not find any information in the docs, if ``get_locs`` should raise when the combination of both keys is not found in MultiIndex. Would prefer that too, but did not want to break anything, hence my question above. Will try to introduce this there

**Comment by phofl** [comment]
This will introduce another case: Should an empty indexer raise or not? Currently the return is an empty DataFrame

File: pandas/core/indexing.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexing.py]: We are raising in ``get_locs`` now. This should be much more stable

File: pandas/tests/indexing/multiindex/test_loc.py [hunk 2]
```diff
@@ -591,6 +591,23 @@ def test_missing_key_raises_keyerror2(self):
...
+            [
+                np.array(["a", "a", "b", "b"]),
+                np.array(["1", "2", "2", "3"]),
+                np.array(["c", "d", "c", "d"]),
+            ],
+            names=["one", "two", "three"],
+        )
+        df = DataFrame(np.random.rand(4, 3), index=mi)
+        msg = r"\('b', '1', slice\(None, None, None\)\)"
+        with pytest.raises(KeyError, match=msg):
```
Comment (jreback) [inline pandas/tests/indexing/multiindex/test_loc.py:606]: can you also add the case where the 3rd level is level out (should also raise)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm ping on green.

File: pandas/tests/indexing/multiindex/test_loc.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/indexing/multiindex/test_loc.py:606]: Done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @jbrockmendel if any comments.

**Comment by jbrockmendel** [comment]
LGTM

Thanks for tackling all these indexing issues @phofl, they can be a real doozy.  FWIW we've done a lot of semi-recent cleanup of the indexing code but MultiIndex has received relatively little attention.  Probably some relatively low-hanging fruit there.

**Comment by phofl** [comment]
Thanks, 

I will have a look for the MultiIndex code paths

**Comment by jbrockmendel** [comment]
@phofl can you merge master, the CI should be fixed

**Comment by phofl** [comment]
Merged master

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**